src/statement.d

```diff
@@ -81,6 +81,33 @@ extern (C++) LabelStatement checkLabeledLoop(Scope* sc, Statement statement)
     return null;
 }
 
+/***********************************************************
+ * Check an assignment is used as a condition.
+ * Intended to be use before the `semantic` call on `e`.
+ * Params:
+ *  e = condition expression which is not yet run semantic analysis.
+ * Returns:
+ *  `e`. If an assignment found, it's replaced with `EqualExp` for better diagnostic.
+ */
+Expression checkAssignmentAsCondition(Expression e)
+{
+    Expression e0;
+    Expression ec = Expression.extractLast(e, &e0);
+    if (ec.op == TOKassign)
+    {
+        e.error("assignment cannot be used as a condition, perhaps == was meant?");
+
+        /* If just return ErrorExp here, no more errors will be reported
+         * even if either of ae operands contains any errors.
+         * By the rewriting, following semantic analysis will report the errors
+         * in ae.e1 and ae.e2.
+         */
```

Comment:
This is not how things must work. The point is not to get all the errors, which is impossible. The point is to mark the node as TOKerror, and mark any node dependent on a TOKerror as also TOKerror. Trying to "repair" the node destroys the efficacy of this.

The point of all this is to remove dependency on checking the global error count - should just check the top level node for TOKerror.